File: ui/tract-renderer/src/main/java/org/cru/godtools/tract/activity/ModalActivity.java
package org.cru.godtools.tract.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.annimon.stream.Optional;

import org.ccci.gto.android.common.util.os.BundleUtils;
import org.cru.godtools.base.model.Event;
import org.cru.godtools.base.tool.activity.ImmersiveActivity;
import org.cru.godtools.base.tool.viewmodel.LatestPublishedManifestDataModel;
import org.cru.godtools.model.Language;
import org.cru.godtools.model.Tool;
import org.cru.godtools.tract.R;
import org.cru.godtools.tract.R2;
import org.cru.godtools.tract.viewmodel.ModalViewHolder;
import org.cru.godtools.xml.model.Manifest;
import org.cru.godtools.xml.model.Modal;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.Locale;

import androidx.annotation.CallSuper;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import butterknife.BindView;

import static androidx.core.app.ActivityOptionsCompat.makeCustomAnimation;
import static org.cru.godtools.base.Constants.EXTRA_LANGUAGE;
import static org.cru.godtools.base.Constants.EXTRA_TOOL;
import static org.cru.godtools.tract.Constants.EXTRA_MANIFEST_FILE_NAME;
import static org.cru.godtools.tract.Constants.EXTRA_MODAL;
import static org.cru.godtools.tract.Constants.EXTRA_PAGE;

public class ModalActivity extends ImmersiveActivity {
    @Nullable
    @BindView(R2.id.modal_root)
    View mModalView;

    @Nullable
    private /*final*/ String mManifestFileName = null;
    @Nullable
    /*final*/ String mTool = Tool.INVALID_CODE;
    @NonNull
    /*final*/ Locale mLocale = Language.INVALID_CODE;
    @Nullable
    private /*final*/ String mPageId;
    @Nullable
    private /*final*/ String mModalId;

    @Nullable
    private Modal mModal;
    @Nullable
    private ModalViewHolder mModalViewHolder;

    public static void start(@NonNull final Activity activity, @NonNull final String manifestFileName,
                             @NonNull final String toolCode, @NonNull final Locale locale, @NonNull final String page,
                             @NonNull final String modal) {
        final Bundle extras = new Bundle(4);
        extras.putString(EXTRA_MANIFEST_FILE_NAME, manifestFileName);
        extras.putString(EXTRA_TOOL, toolCode);
        BundleUtils.putLocale(extras, EXTRA_LANGUAGE, locale);
        extras.putString(EXTRA_PAGE, page);
        extras.putString(EXTRA_MODAL, modal);

        ContextCompat.startActivity(activity, new Intent(activity, ModalActivity.class).putExtras(extras),
                                    makeCustomAnimation(activity, R.anim.activity_fade_in, R.anim.activity_fade_out)
                                            .toBundle());
    }

    public ModalActivity() {
        super(true, R.layout.activity_modal);
    }

    /* BEGIN lifecycle */

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Intent intent = getIntent();
        final Bundle extras = intent != null ? intent.getExtras() : null;
        if (extras != null) {
            mManifestFileName = extras.getString(EXTRA_MANIFEST_FILE_NAME, mManifestFileName);
            mTool = extras.getString(EXTRA_TOOL, mTool);
            mLocale = BundleUtils.getLocale(extras, EXTRA_LANGUAGE, mLocale);
            mPageId = extras.getString(EXTRA_PAGE, mPageId);
            mModalId = extras.getString(EXTRA_MODAL, mModalId);
        }

        // finish now if this activity is in an invalid state
        if (!validStartState()) {
            finish();
            return;
        }

        setupDataModel();
        startLoaders();
    }

    @Override
    @CallSuper
    public void onContentChanged() {
        super.onContentChanged();
        setupModalViewHolder();
    }

    @Override
    protected void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onContentEvent(@NonNull final Event event) {
        checkForDismissEvent(event);
    }

    @Override
    protected void onPause() {
        super.onPause();
        EventBus.getDefault().unregister(this);
    }

    /* END lifecycle */

    private boolean validStartState() {
        return mTool != null;
    }

    // region Data Model
    private LatestPublishedManifestDataModel mDataModel;

    private void setupDataModel() {
        mDataModel = (new ViewModelProvider(this)).get(LatestPublishedManifestDataModel.class);
        mDataModel.getToolCode().setValue(mTool);
        mDataModel.getLocale().setValue(mLocale);
    }
    // endregion Data Model

    private void startLoaders() {
        mDataModel.getManifest().observe(this, this::updateModal);
    }

    void updateModal(@Nullable final Manifest manifest) {
        mModal = Optional.ofNullable(manifest)
                .map(m -> m.findPage(mPageId))
                .map(p -> p.findModal(mModalId))
                .orElse(null);
        if (mModal == null) {
            finish();
        }
        updateModalViewHolder();
    }

    private void setupModalViewHolder() {
        if (mModalView != null) {
            mModalViewHolder = ModalViewHolder.forView(mModalView);
            updateModalViewHolder();
        }
    }

    private void updateModalViewHolder() {
        if (mModalViewHolder != null) {
            mModalViewHolder.bind(mModal);
        }
    }

    private void checkForDismissEvent(@NonNull final Event event) {
        if (mModal != null) {
            if (mModal.getDismissListeners().contains(event.id)) {
                finish();
            }
        }
    }
}

File: ui/tract-renderer/src/main/kotlin/org/cru/godtools/tract/activity/ModalActivity.kt
package org.cru.godtools.tract.activity

import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.annotation.MainThread
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.distinctUntilChanged
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.lifecycle.HiltViewModel
import io.fluidsonic.locale.toPlatform
import javax.inject.Inject
import org.ccci.gto.android.common.androidx.lifecycle.combineWith
import org.ccci.gto.android.common.util.os.getLocale
import org.ccci.gto.android.common.util.os.putLocale
import org.cru.godtools.base.EXTRA_LANGUAGE
import org.cru.godtools.base.EXTRA_PAGE
import org.cru.godtools.base.EXTRA_TOOL
import org.cru.godtools.base.tool.model.Event
import org.cru.godtools.base.tool.service.ManifestManager
import org.cru.godtools.base.tool.viewmodel.LatestPublishedManifestDataModel
import org.cru.godtools.base.tool.viewmodel.ToolStateHolder
import org.cru.godtools.base.ui.activity.BaseBindingActivity
import org.cru.godtools.shared.tool.parser.model.tract.Modal
import org.cru.godtools.shared.tool.parser.model.tract.TractPage
import org.cru.godtools.tool.tract.R
import org.cru.godtools.tool.tract.databinding.TractModalActivityBinding
import org.cru.godtools.tract.EXTRA_MODAL
import org.cru.godtools.tract.ui.controller.ModalController
import org.cru.godtools.tract.ui.controller.bindController
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

internal fun Activity.startModalActivity(modal: Modal) = startActivity(
    Intent(this, ModalActivity::class.java).putExtras(
        Bundle(4).apply {
            putString(EXTRA_TOOL, modal.manifest.code)
            putLocale(EXTRA_LANGUAGE, modal.manifest.locale?.toPlatform())
            putString(EXTRA_PAGE, modal.page.id)
            putString(EXTRA_MODAL, modal.id)
        }
    ),
    ActivityOptions.makeCustomAnimation(
        this,
        org.cru.godtools.ui.R.anim.activity_fade_in,
        org.cru.godtools.ui.R.anim.activity_fade_out
    ).toBundle()
)

@AndroidEntryPoint
class ModalActivity : BaseBindingActivity<TractModalActivityBinding>(R.layout.tract_modal_activity) {
    private val dataModel: ModalActivityDataModel by viewModels()
    private val toolState: ToolStateHolder by viewModels()
    @Inject
    internal lateinit var modalControllerFactory: ModalController.Factory

    // region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        intent?.extras?.let { extras ->
            dataModel.toolCode.value = extras.getString(EXTRA_TOOL)
            dataModel.locale.value = extras.getLocale(EXTRA_LANGUAGE)
            dataModel.pageId.value = extras.getString(EXTRA_PAGE)
            dataModel.modalId.value = extras.getString(EXTRA_MODAL)
        }

        // finish if this activity is in an invalid state
        if (!validStartState()) {
            finish()
            return
        }
        dataModel.modal.observe(this) { if (it == null) finish() }
    }

    override fun onBindingChanged() {
        binding.modalLayout.bindController(modalControllerFactory, this, toolState.toolState)
            .also { dataModel.modal.observe(this, it) }
    }

    override fun onResume() {
        super.onResume()
        eventBus.register(this)
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onContentEvent(event: Event) {
        if (event.tool != dataModel.toolCode.value && event.locale != dataModel.locale.value) return
        checkForDismissEvent(event)
    }

    override fun onPause() {
        super.onPause()
        eventBus.unregister(this)
    }
    // endregion Lifecycle

    override val toolbar: Toolbar? get() = null

    private fun validStartState() = dataModel.toolCode.value != null

    private fun checkForDismissEvent(event: Event) {
        if (dataModel.modal.value?.dismissListeners?.contains(event.id) == true) finish()
    }
}

@HiltViewModel
class ModalActivityDataModel @Inject constructor(manifestManager: ManifestManager) :
    LatestPublishedManifestDataModel(manifestManager) {
    val pageId = MutableLiveData<String?>()
    val modalId = MutableLiveData<String?>()

    val modal =
        manifest.combineWith(pageId.distinctUntilChanged(), modalId.distinctUntilChanged()) { manifest, page, modal ->
            (manifest?.findPage(page) as? TractPage)?.findModal(modal)
        }
}
